JAVA-1286: Support setting and retrieving udt fields in QueryBuilder

tolbertam opened the pull request:
For [JAVA-1286](https://datastax-oss.atlassian.net/browse/JAVA-1286).

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/querybuilder/Utils.java:
```diff
@@ -32,7 +32,7 @@
 abstract class Utils {
 
-    private static final Pattern cnamePattern = Pattern.compile("\\w+(?:\\[.+\\])?");
+    private static final Pattern cnamePattern = Pattern.compile("\\w+(?:\\[.+\\]|\\.\\w+)?");
```
I admit that this may not be the best route.  Also supporting this may break existing user code (i.e. if they have columns named `"myc.ol"`).

tolbertam commented on driver-core/src/test/java/com/datastax/driver/core/querybuilder/QueryBuilderUDTExecutionTest.java:
```diff
@@ -87,4 +88,39 @@ public void should_handle_collections_of_UDT() throws Exception {
+        assertThat(userType).isNotNull();
+
+        session().execute(createTable(table).addPartitionKey("k", DataType.text())
+                .addUDTColumn("u", udtLiteral(udt))
```
Maybe we should consider renaming `udtLiteral` to something else since unfrozen udts are now allowed (although maybe in future version?)

olim7t replied:
The name refers to expressing the type with a string literal, I think this still holds even if the type is unfrozen.

adutra commented:
To be honest I would be in favor of a different approach: allow `QueryBuilder.raw()` (or maybe `QueryBuilder.column()`) to be used as a column name in a set assignment. 

This would lead to a change (hopefully backwards-compatible) in the signature of `set()`: `public static Assignment set(Object name, Object value)`, and would require some changes in the `Assignment` class to detect when a variable name should or should not be enclosed in single quotes.

But I think this would be ultimately more in line with the way we currently protect strings from being mistankenly quoted by the query builder.

tolbertam commented:
> To be honest I would be in favor of a different approach: allow QueryBuilder.raw() (or maybe QueryBuilder.column()) to be used as a column name in a set assignment.

I am on the fence with that, but I think I've come around on your suggestion.  On the one hand, the current proposed code can break some users if they for whatever reason had periods in their column names and weren't quoting column names theirselves.   On the other hand, users should be in the right to expect the driver to not mistakenly quote the name in this particular case.

I think quoting the name for the user as it currently exists is inconsistent with other parts of our API and we should consider changing this in the future.  For example, we won't quote the keyspace when passed into `Cluster.connect(keyspace)`.  However, we can't simply change that here.  And even changing that in a future major release could be a problem too for users (even if we are really careful about communicating it).

In that case I agree that the right approach then is to come up with an alternative way to say "Here is my name, don't touch it" (as you've proposed) by allowing use of `column` or `raw` with `set`.   There will still be users who encounter their name being escaped when they wouldn't expect, but at least now they'll have a workaround.

newkek commented:
+1 for the `QueryBuilder.raw()` approach + we should study where else we could allow it to give people escape plans when the driver doesn't behave as expected.

olim7t commented:
Changed base to 3.x, make sure you use that branch for future rebases.

tolbertam committed:
JAVA-1286: Support setting and retrieving udt fields in QueryBuilder

CASSANDRA-7423 in Apache Cassandra 3.6 adds the ability to update
individual UDT fields.

adutra commented:
Rebased on 3.x. 

On a side note, do we have an agreement on modifying `cnamePattern` vs modifying `set` signature?

adutra commented:
Turns out changing the parameter of a public method _is_ a breaking change :-/ it's a pity because this pattern could be applied in lots of places. But anyway, semver says no :)

olim7t commented on driver-core/src/main/java/com/datastax/driver/core/querybuilder/Utils.java:
```diff
@@ -33,7 +33,7 @@
 
     private static final Pattern alphanumeric = Pattern.compile("\\w+"); // this includes _
-    private static final Pattern cnamePattern = Pattern.compile("\\w+(?:\\[.+\\])?");
+    private static final Pattern cnamePattern = Pattern.compile("\\w+(?:\\[.+\\]|\\.\\w+)?");
```
This does not handle "a.b.c" if there are multiple levels of nesting (is it possible?)

olim7t commented:
We could get around the breaking change by leaving `set(String, Object)` and adding an overload `set(Object, Object)`.
For selection, there is already a workaround with `select().raw("a.b")`.

Maybe we should also have a programmatic way to express it (`field("a", "b")`?), for cases when you generate a query from a UDT definition.

adutra commented:
> We could get around the breaking change by leaving set(String, Object) and adding an overload set(Object, Object).

Indeed, done.

> Maybe we should also have a programmatic way to express it (field("a", "b")?), for cases when you generate a query from a UDT definition.

Agreed and done. Although I named the method `path` instead of `field`, as I sense that in the future that dotted notation could be useful in more places than are currently supported.

tolbertam commented:
The changes look good to me, can't approve since I created the PR, but this has my thumbs up.

olim7t commented on driver-core/src/main/java/com/datastax/driver/core/querybuilder/QueryBuilder.java:
```diff
@@ -589,6 +589,35 @@ public static Assignment set(String name, Object value) {
+     * <li>{@link #column(String)}: {@code set(column("name"), value)}
+     * generates {@code "name" = value}.</li>
+     * <li>{@link #path(String...)}: {@code set(path("udt", "field"), value)}
+     * generates {@code name.field = value}.</li>
```
I wonder if it's worth mentioning `column`. I don't see any reason why you would want to use it here, and it's more confusing than helpful.

olim7t reviewed (approved):
LGTM, I did a few changes to the javadoc to address my remark.
